### PR TITLE
Migrate database from `firestore` to `mongodb`

### PR DESCRIPTION
**Pull request by amankumarsingh77**
### Summary

Migrate all the services and database from `firestore` to `mongodb`.

### Why migrating?
- Firebase requires numerous environment variables, which can complicate Lambda deployments. 
- After conducting stress tests, it was found that MongoDB outperforms Firestore in terms of performance. 
- For projects on the free plan, the read and write limits of MongoDB offer significant advantages over those of Firestore.

### How to setup mongoDB?
1. **Create a [MongoDB Atlas](https://www.mongodb.com/cloud/atlas/register) Account (Free M0 Cluster)**:
    - Limitations: 
       - 100 (read/write)/per second.
       - 500 connections.
       - 500 collections.
       - 3 indexes.
       - 512 mb storage.
       - 10 GB in and 10 GB out data transferred per period (7days).

2. **Create a Cluster**:
    - Once logged in, click on `Create a Cluster`.
    - Choose the free tier `M0` and configure your cluster as needed.
    - Click `Create Deployment`. This may take a few minutes.

3. **Create a Database User**:
    - In your cluster view, go to the `Database Access` tab.
    - Click `+ Add New Database User`.
    - Create a new user with the appropriate roles and a strong password.

4. **Configure Network Access**:
    - Go to the `Network Access` tab.
    - Click `+ Add IP Address`.
    - Add your IP address or allow access from anywhere (`0.0.0.0/0`), though this is not recommended for production.

5. **Get the Connection String**:
    - In your cluster view, click `Connect`.
    - Select driver as Node.js and version as 5.5 or later
    - Copy the connection string. It should look something like this:
      ```
      mongodb+srv://<username>:<password>@cluster0.mongodb.net/<dbname>?retryWrites=true&w=majority
      ```

6. **Set Up Environment Variable**:
    - Create a `.env` file in the root of your project.
    - Add your MongoDB connection string to the `.env` file:
      ```env
      MONGO_URL=mongodb+srv://<username>:<password>@cluster0.mongodb.net/<dbname>?retryWrites=true&w=majority
      ```

### Changes:
- Initially, read and write operations in Firestore were costly, leading to the monthly quota being reached and causing our API endpoints that relied on Firestore to fail.
- After testing several databases, MongoDB proved to be the best fit for our requirements.
- All services that previously used Firestore were rewritten to be compatible with MongoDB, along with their respective tests.
- All Firestore and Firebase dependencies were removed from the codebase.

### Documentation changes:
- `README` file has been updated to guide devs to setup `mongodb` for local development.
- `.env.example` has been updated with new mongodb env variables as per the requirement.

### Migration Script (not part of this code):
- Script will be provided separately to migrate data from `firestore` collections to `mongodb` collections.

### ScreenShots:

Screenshot 2024-07-25 at 8 58 57 PM

https://github.com/user-attachments/assets/79d6e0c0-2514-4733-a6bf-d04566bc3a28

## Checklist

- [x] I have tested the changes locally and they work as expected.
- [x] I have added/updated tests that cover the changes.
- [x] I have updated the documentation to reflect the changes.
- [x] This pull request follows the project's coding guidelines.

### CODE REVIEW

File: .github/workflows/pre-commit.yml
Comment (aps08) [inline .github/workflows/pre-commit.yml:1]: No changes should be here

File: server/.env_example
Comment (aps08) [inline server/.env_example:1]: UNIVERSE_DOMAIN domain is used for firebase,  not required.

File: server/utils/sendEmail.js
Comment (aps08) [inline server/utils/sendEmail.js:1]: remove comment

File: server/utils/scripts/envSchema.js
Comment (aps08) [inline server/utils/scripts/envSchema.js:1]: remove UNIVERSE_DOMAIN

File: server/controllers/auth/verify.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ async function verifyTokenController(req, res, next) {
         statusCode: 403,
       });
 
-    const user = await fetchUserFromId(userToken.userId);
+    const user = await fetchUserFromId(userToken.user);
```
Comment (Pratiyushkumar) [inline server/controllers/auth/verify.js:35]: wont it be _id instead of user?

**Review by Pratiyushkumar — COMMENTED** [review]
The doubt I have asked in comment

File: server/controllers/auth/verify.js (hunk 1 above)
Comment (amankumarsingh77) [reply server/controllers/auth/verify.js:35]: No. As of now we storing the user ref in UserToken thats why we use `user` instead of passing `_id`.

File: jest-mongodb-config.js
Comment (aps08) [inline jest-mongodb-config.js:1]: This file should be at root or inside the server folder ?

File: jest.config.js
Comment (aps08) [inline jest.config.js:1]: This file should be at root or inside the server folder ?

File: server/__tests__/unit/services/ContactUs.js [hunk 2]
```diff
@@ -49,32 +75,26 @@ describe("createForm", () => {
...
-    expect(result).toBeNull();
+    await expect(createForm(formData)).rejects.toThrow("MongoDB operation failed");
   });
 
-  it("should return a ContactUs instance with default values when Firestore set operation returns partial data (edge case)", async () => {
-    jest.spyOn(ContactUsCollection, "doc").mockReturnValue({
-      set: jest.fn().mockResolvedValue({})
-    });
+  it("should return a ContactUs instance with default values when MongoDB save operation returns partial data (edge case)", async () => {
+    // jest.spyOn(ContactUs.prototype, "save").mockResolvedValue({});
```
Comment (aps08) [inline server/__tests__/unit/services/ContactUs.js]: Why this line is commented out ?

File: server/__tests__/unit/services/Keys.js [hunk 3]
```diff
@@ -1,67 +1,85 @@
...
-}));
+beforeAll(async () => {
+  const mongoServer = await MongoMemoryServer.create();
+  const mongo_uri = mongoServer.getUri();
+  await mongoose.connect(mongo_uri);
+});
 
-afterEach(() => {
-  jest.restoreAllMocks();
+afterAll(async () => {
```
Comment (aps08) [inline server/__tests__/unit/services/Keys.js:13]: Is this required here ? Some test files doesn't have `afterAll` and `beforeAll`.

File: server/__tests__/unit/services/Subscriptions.js [hunk 4]
```diff
@@ -1,50 +1,78 @@
...
   fetchSubscriptionByuserid,
 } = require("../../../services");
-
-const { mockUsers } = require("../../../utils/mocks/Users");
-const { SubscriptionCollection } = require("../../../utils/firestore");
+const { mockSubscriptions } = require("../../../utils/mocks/Subscriptions");
 const { Subscriptions } = require("../../../models");
+const { MongoMemoryServer } = require("mongodb-memory-server");
+
+beforeAll(async () => {
```
Comment (aps08) [inline server/__tests__/unit/services/Subscriptions.js:10]: Is this required here ? Some test files doesn't have `afterAll` and `beforeAll`.

File: server/__tests__/unit/utils/scripts/envSchema.js [hunk 5]
```diff
@@ -17,16 +17,7 @@ const validEnv = {
...
-  FIRESTORE_PRIVATE_KEY: "randomText",
-  FIRESTORE_PRIVATE_ID: "randomText",
-  FIRESTORE_CLIENT_EMAIL: "randomText",
-  FIRESTORE_CLIENT_ID: "randomText",
-  FIRESTORE_TOKEN_URI: "randomText",
-  FIRESTORE_AUTH_PROVIDER: "randomText",
-  FIRESTORE_CLIENT_CERT: "randomText",
-  UNIVERSE_DOMAIN: "randomText",
-  FIRESTORE_AUTH_URI: "randomText",
+  UNIVERSE_DOMAIN: "randomText"
```
Comment (aps08) [inline server/__tests__/unit/utils/scripts/envSchema.js]: `UNISERSE_DOMAIN` not required.

File: server/models/Users.js [hunk 6]
```diff
@@ -1,110 +1,104 @@
...
-  get data() {
-    return {
-      firstName: this.firstName,
-      lastName: this.lastName,
-      email: this.email,
-      userId: this.userId,
-      userType: this.userType,
-    };
-  }
+// Pre-save hook to hash the password before saving
```
Comment (aps08) [inline server/models/Users.js]: Why this code is commented out ?

File: server/services/Images.js [hunk 7]
```diff
@@ -11,7 +12,7 @@ const s3 = new S3Client({
   },
 });
 
-async function uploadToS3(file, imageName, extension) {
+const uploadToS3 = async(file, imageName, extension) => {
```
Comment (aps08) [inline server/services/Images.js]: Why this change ? from `async function` to `const` ? Change it to the previous format. Make sure all the services have same format.

File: server/services/Users.js [hunk 8]
```diff
@@ -1,60 +1,43 @@
...
-
+const { Users, Subscriptions, Keys } = require("../models");
+const mongoose = require("mongoose");
 /**
  * Checks if provided email exists in the user collections
  * @param {string} email
  * @returns {Promise<boolean>} true if email already exists and else false
  **/
-async function emailRecordExists(email) {
+const emailRecordExists = async (email) => {
```
Comment (aps08) [inline server/services/Users.js]: Again change in the function style. I would like to know why ?

**Comment by amankumarsingh77** [comment]
@aps08 We have fixed all the minor issues which were mentioned above in the review.

File: jest-mongodb-config.js
Comment (amankumarsingh77) [reply jest-mongodb-config.js:1]: Yeah. We need to have it in the root directory as we are running our test command in the root directory. I have tried putting them in the /server and it throws an error.

File: jest.config.js
Comment (amankumarsingh77) [reply jest.config.js:1]: Config files must always be in root directory or it does not work.

**Review by aps08 — APPROVED** [review]

**Comment by aps08** [comment]
Will be merged tomorrow in the meeting.
